workspaces/copilot/plugins/copilot-backend/src/utils/CopilotCredentialsProvider.ts

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright 2024 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Config } from '@backstage/config';
+import {
+  DefaultGithubCredentialsProvider,
+  GithubCredentials,
+  ScmIntegrations,
+} from '@backstage/integration';
+
+/**
+ * Information required to access the GitHub API
+ *
+ * @public
+ */
+export type GithubInfo = {
+  credentials: GithubCredentials;
+  apiBaseUrl: string;
+  enterprise: string;
+};
+
+/**
+ * Interface for providing credentials for accessing the copilot API
+ *
+ * @public
+ */
+export interface CopilotCredentialsProvider {
+  /**
+   * Retrieve the credentials required to access the copilot API
+   *
+   * @public
+   */
+  getCredentials(): Promise<GithubInfo>;
+}
+
+export class DefaultCopilotCredentialsProvider
+  implements CopilotCredentialsProvider
+{
+  private readonly host: string;
+  private readonly enterprise: string;
+  private readonly integrations: ScmIntegrations;
+  private readonly credentialsProvider: DefaultGithubCredentialsProvider;
+
+  constructor(options: { config: Config }) {
+    const { config } = options;
+
+    this.integrations = ScmIntegrations.fromConfig(config);
+    this.credentialsProvider =
+      DefaultGithubCredentialsProvider.fromIntegrations(this.integrations);
+
+    this.host = config.getString('copilot.host');
+    this.enterprise = config.getString('copilot.enterprise');
+
+    if (!this.host) {
+      throw new Error('The host configuration is missing from the config.');
+    }
+
+    if (!this.enterprise) {
+      throw new Error(
+        'The enterprise configuration is missing from the config.',
+      );
+    }
+  }
+
+  async getCredentials(): Promise<GithubInfo> {
+    const githubConfig = this.integrations.github.byHost(this.host)?.config;
+
+    if (!githubConfig) {
+      throw new Error(
+        `GitHub configuration for host "${this.host}" is missing or incomplete.`,
+      );
+    }
+
+    const apiBaseUrl = githubConfig.apiBaseUrl ?? 'https://api.github.com';
+
+    const credentials = await this.credentialsProvider.getCredentials({
+      url: apiBaseUrl,
+    });
```

Comment:
I think this needs to change to work for most people.

For most users of github they have `github.com` as the host in their integrations config. The code here means that even if they configure github.com in the copilot section it will try to resolve api.github.com and fail (most people would not have this configured in their config)

So i thin kthis needs to be `url: host` to make it work for most people.

There is also the issue that the integrations config can be done with a github app which would not have access to the enterprise API at all. This is how we have it configured and probably most users of GH will use an app (I'm guessing).

I think that the default here could be to have a specific token in the copilot section as this token is specific to accessing the enterprise API.